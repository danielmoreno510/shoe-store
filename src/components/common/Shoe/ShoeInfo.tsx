import React from 'react';
import {View, Text} from 'react-native';
import {Icon} from 'native-base';
import RNPickerSelect from 'react-native-picker-select';
import {filter, find} from 'lodash';

import Button from '../Button';
import styles, {pickerSelectStyles} from './styles';

const ShoeInfo: React.FC<ShoeProps> = ({
  shoe,
  cart,
  addToCart,
  showProduct,
  twoSection,
  oneSection,
  newSize,
}) => {
  const [size, setSize] = React.useState<number>(4.5);
  const getQuantity = () => {
    const quantityShoes = filter(cart, ['id', shoe.id]);
    const shoesAvailable = find(shoe.sizes, ['value', size]);
    return shoesAvailable.quantity - quantityShoes.length;
  };

  return (
    <View style={[twoSection && styles.infoContainer, oneSection && styles.oneInfoContainer]}>
      <View style={styles.sectionContainer}>
        <Text style={styles.reference}>{shoe.reference}</Text>
        <Text style={styles.brand}>{shoe.brand}</Text>
        <View style={styles.sizeContainer}>
          <Text style={styles.size}>Size: </Text>
          {!newSize ? (
            <View style={styles.quantityContainer}>
              <View style={styles.sizeSelectContainer}>
                <RNPickerSelect
                  onValueChange={value => setSize(value)}
                  items={shoe.sizes}
                  value={size}
                  style={pickerSelectStyles}
                />
              </View>
              <Text style={styles.quantity}>{` ${getQuantity()}U`}</Text>
            </View>
          ) : (
            <Text style={styles.size}>{newSize}</Text>
          )}
        </View>
      </View>
      <View style={[styles.sectionContainer]}>
        <Text style={styles.price}>{`$${shoe.price}`}</Text>
        <View style={styles.buttonsContainer}>
          {addToCart && getQuantity() > 0 && (
            <Button testID="addToCart" onPress={() => addToCart({...shoe, size})} styleButton={styles.button}>
              <Icon type="MaterialIcons" name="shopping-cart" style={styles.icon} />
            </Button>
          )}
          {showProduct && (
            <Button testID="showProduct" onPress={() => showProduct(shoe)} styleButton={styles.button}>
              <Icon type="MaterialIcons" name="visibility" style={styles.icon} />
            </Button>
          )}
        </View>
        {!oneSection && <Text>{`Date: ${shoe.date}`}</Text>}
      </View>
    </View>
  );
};

export default ShoeInfo;
