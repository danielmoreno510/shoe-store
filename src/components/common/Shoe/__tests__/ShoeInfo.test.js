import React from 'react';
import {shallow} from 'enzyme';

import ShoeInfo from '../ShoeInfo';

describe('ShoeInfo component', () => {
  describe('GIVEN a set of valid props', () => {
    const testProps = {
      shoe: {
        id: 3,
        brand: 'Nike',
        reference: 'Blazer Low Platform',
        price: 200,
        sizes: [
          {label: '4.5', value: 4.5, quantity: 1},
          {label: '5', value: 5, quantity: 2},
          {label: '5.5', value: 5.5, quantity: 9},
          {label: '6', value: 6, quantity: 7},
          {label: '6.5', value: 6.5, quantity: 5},
          {label: '7', value: 7, quantity: 4},
        ],
        images: ['blazerLowPlatform1', 'blazerLowPlatform2', 'blazerLowPlatform3', 'blazerLowPlatform4'],
        date: '2021-06-01',
      },
      cart: [],
      addToCart: jest.fn(),
      showProduct: jest.fn(),
      twoSection: true,
      oneSection: false,
    };

    describe('WHEN the component is rendered', () => {
      const shallowWrapper = shallow(<ShoeInfo {...testProps} />);

      it('THEN should display a regular ShoeInfo', () => {
        expect(shallowWrapper).toMatchSnapshot();
      });
    });

    describe('WHEN the addToCart showProduct button are pressed', () => {
      const customProps = {...testProps, oneSection: true, newSize: 5};
      const shallowWrapper = shallow(<ShoeInfo {...customProps} />);
      shallowWrapper.find({testID: 'addToCart'}).simulate('press');
      shallowWrapper.find({testID: 'showProduct'}).simulate('press');

      it('THEN should addToCart and showProduct should be called', () => {
        expect(customProps.addToCart).toBeCalledTimes(1);
        expect(customProps.showProduct).toBeCalledTimes(1);
      });
    });
  });
});
