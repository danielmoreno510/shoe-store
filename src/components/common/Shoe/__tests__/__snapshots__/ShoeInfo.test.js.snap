// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoeInfo component GIVEN a set of valid props WHEN the component is rendered THEN should display a regular ShoeInfo 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      false,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 22,
          "fontWeight": "500",
        }
      }
    >
      Blazer Low Platform
    </Text>
    <Text
      style={
        Object {
          "fontSize": 18,
        }
      }
    >
      Nike
    </Text>
    <View
      style={
        Object {
          "alignContent": "center",
          "flexDirection": "row",
          "marginTop": 12,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 20,
          }
        }
      >
        Size: 
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "borderColor": "#f1b30a",
              "borderRadius": 4,
              "borderWidth": 1,
              "width": 40,
            }
          }
        >
          <mockConstructor
            items={
              Array [
                Object {
                  "label": "4.5",
                  "quantity": 1,
                  "value": 4.5,
                },
                Object {
                  "label": "5",
                  "quantity": 2,
                  "value": 5,
                },
                Object {
                  "label": "5.5",
                  "quantity": 9,
                  "value": 5.5,
                },
                Object {
                  "label": "6",
                  "quantity": 7,
                  "value": 6,
                },
                Object {
                  "label": "6.5",
                  "quantity": 5,
                  "value": 6.5,
                },
                Object {
                  "label": "7",
                  "quantity": 4,
                  "value": 7,
                },
              ]
            }
            onValueChange={[Function]}
            style={
              Object {
                "inputAndroid": Object {
                  "fontSize": 20,
                  "fontWeight": "bold",
                  "textAlign": "center",
                },
                "inputIOS": Object {
                  "fontSize": 20,
                  "fontWeight": "bold",
                  "textAlign": "center",
                },
              }
            }
            value={4.5}
          />
        </View>
        <Text
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
           1U
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flex": 1,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#f1b30a",
          "fontSize": 28,
          "fontWeight": "bold",
          "marginTop": 12,
        }
      }
    >
      $200
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "paddingTop": 12,
        }
      }
    >
      <CustomButton
        onPress={[Function]}
        styleButton={
          Object {
            "margin": 12,
            "width": 54,
          }
        }
        testID="addToCart"
      >
        <Component
          name="shopping-cart"
          style={
            Object {
              "color": "white",
            }
          }
          type="MaterialIcons"
        />
      </CustomButton>
      <CustomButton
        onPress={[Function]}
        styleButton={
          Object {
            "margin": 12,
            "width": 54,
          }
        }
        testID="showProduct"
      >
        <Component
          name="visibility"
          style={
            Object {
              "color": "white",
            }
          }
          type="MaterialIcons"
        />
      </CustomButton>
    </View>
    <Text>
      Date: 2021-06-01
    </Text>
  </View>
</View>
`;
